### ISSUE TITLE
In place binary upgrades for the master

### ISSUE THREAD

**Issue by roberthbailey** [kubernetes/kubernetes#8151]
#6075 describes how we plan to do master upgrades on GCE and GKE where we can rely on having a persistent disk to store cluster configuration and etcd data. We should also have a way to upgrade masters in place (flip the master binaries) without blowing away the whole node. 

The tentative proposal is to follow the kube-push model and write a new manifest file so that the kubelet will launch new versions of the master pods. 

For GCE & GKE we have some itinerant Salt gunk we still haven't gotten away from, so there might be master config aspects that need to be handled as well. 

@zmerlynn @davidopp

**Comment by davidopp** [comment]
I guess the key point here is that this would enable master upgrade for platforms without PDs, for example bare metal. (And, if AWS/Azure if we store the etcd data locally on those platforms, but maybe it would be better to just have etcd write to their equivalent of persistent disks if that's the case?)

**Comment by dchen1107** [comment]
cc/ @dchen1107

**Comment by dchen1107** [comment]
Most of work are already done through #5011, but we are missing: 

1) Our release procedure should publish our kube-* images with proper release tag to gcr.io
2) kubectl should have command supported for cluster upgrade, especially for master components by taking argument of which release version.

**Comment by davidopp** [comment]
My understanding is that the Warsaw team offered to work on this and the "binary upgrade" piece of #6079 (ignore that it says post-1.0; we need it for 1.0). In other words, all the pieces we still need for OSS master and node upgrade.

@rsokolowski @wojtek-t @fgrzadkowski @gmarek @piosz @jszczepkowski @marekbiskup

**Comment by dchen1107** [comment]
@davidopp We had meeting with Warsaw folks yesterday talking about this. Thanks!

**Comment by davidopp** [comment]
@piosz now that #9063 has been merged, can we move this issue out of the 1.0 milestone? Or do you think the feature needs additional testing?

**Comment by davidopp** [comment]
Oh, I guess in this comment
https://github.com/GoogleCloudPlatform/kubernetes/issues/8082#issuecomment-107650747
you said you thought we should test this as part of the cluster upgrade script?

**Comment by davidopp** [comment]
@saad-ali added it to the test, so I think we're only waiting on a fix to #9110 (which @dchen1107 is working on) and then we can move this out of the 1.0 milestone.

**Comment by piosz** [comment]
#9110 fixed so moving this to post-1.0 milestone.

**Comment by davidopp** [comment]
Thanks @piosz and @dchen1107 !

**Comment by davidopp** [comment]
(BTW it would probably be good to do some more testing on this feature.)

**Comment by gmarek** [comment]
@davidopp @roberthbailey @dchen1107 is this already fixed?

**Comment by davidopp** [comment]
I think so. The last TODO was to have a test, and I believe we're testing it
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/cluster_upgrade.go#L247
